alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+            }
+        }
+
+        for index in self.index..next_index {
+            self.terminal_damage[index].reset();
+        }
+
+        // Update information for the next iteration.
+        self.rect = next_rect;
+        self.index = next_index;
+
+        Some(self.overdamage(rect))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use crate::config::UiConfig;
+    use alacritty_terminal::grid::Dimensions;
+    use alacritty_terminal::term::{
+        DamagedLine, RenderableCursor as TermRenderableCursor, SizeInfo, Term,
+    };
+
+    #[test]
+    fn damage_fully_damaged() {
+        let size = SizeInfo::new(10., 10., 1., 1., 0., 0., false);
+        let mut term = Term::new(&UiConfig::default().terminal_config, size, ());
+        let size: SizeInfo<u32> = size.into();
+
+        let cursor_point = term.grid().cursor.point;
+        let cursor = TermRenderableCursor { point: cursor_point, shape: Default::default() };
+
+        term.damage().mark_fully_damaged();
+        let mut renderer_damage = RendererDamage::new(term.damage(), size, cursor, None, None, 0);
+
+        assert_eq!(
+            renderer_damage.next(),
+            Some(Rect { x: 0, y: 0, width: size.width(), height: size.height() })
+        );
+        assert_eq!(renderer_damage.next(), None);
+    }
+
+    #[test]
+    fn damage_just_cursor() {
+        let size = SizeInfo::new(10., 10., 1., 1., 0., 0., false);
+        let mut term = Term::new(&UiConfig::default().terminal_config, size, ());
+        let size: SizeInfo<u32> = size.into();
+        let num_cols = term.columns();
+
+        let cursor_point = term.grid().cursor.point;
+        let cursor = TermRenderableCursor { point: cursor_point, shape: Default::default() };
+
+        let term_damage = term.damage();
+        // Ensure it's not damaged.
+        term_damage.take_fully_damaged();
+        let mut renderer_damage = RendererDamage::new(term_damage, size, cursor, None, None, 0);
+
+        let mut cursor_line = DamagedLine::undamaged(num_cols);
+        cursor_line.expand(cursor_point.column.0, cursor_point.column.0);
+        let cursor_rect =
+            RendererDamage::rect_for_line(&size, cursor_point.line.0 as usize, cursor_line);
+        let cursor_rect = renderer_damage.overdamage(cursor_rect);
+
+        assert_eq!(renderer_damage.next(), Some(cursor_rect));
+        assert_eq!(renderer_damage.next(), None);
+    }
+
+    #[test]
+    fn damage_shaping() {
+        let size = SizeInfo::new(10., 100., 1., 1., 0., 0., false);
+        let mut term = Term::new(&UiConfig::default().terminal_config, size, ());
+        let size: SizeInfo<u32> = size.into();
+        let num_cols = term.columns();
+
+        let cursor_point = term.grid().cursor.point;
+        let cursor = TermRenderableCursor { point: cursor_point, shape: Default::default() };
+
+        let term_damage = term.damage();
+        // Ensure it's not damaged.
+        term_damage.take_fully_damaged();
```

Comment:
This confuses me, why do we need to ensure it's not damaged after we *just* started the test? This also applies to your other tests.

Is a newly created terminal fully damaged?

Why is the comment `ensure`, when you're not checking anything? If anything that should say "clear existing damage" or something.

Reply:
Yeah, newly created terminal is always fully damaged.

Reply:
Should update the comment accordingly then.